File: python/stack/implement-stack-using-queues.py
class MyStack:

    def __init__(self):
        self.q1 = deque()
        self.q2 = deque()

    def push(self, x: int) -> None:
        self.q1.append(x)

    def pop(self) -> int:
        while len(self.q1) > 1:
            self.q2.append(self.q1.popleft())

        popped = self.q1.popleft()
        while self.q2:
            self.q1.append(self.q2.popleft())

        return popped

    def top(self) -> int:
        while self.q1:
            top = self.q1.popleft()
            self.q2.append(top)

        while self.q2:
            self.q1.append(self.q2.popleft())

        return top

    def empty(self) -> bool:
        return len(self.q1) == 0

# Your MyStack object will be instantiated and called as such:
# obj = MyStack()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.top()
# param_4 = obj.empty()

File: c++/stack/implement-stack-using-queues.cc
class MyStack {
public:
    MyStack() {

    }

    void push(int x) {
        queue_1_.push(x);
    }

    int pop() {
        while (queue_1_.size() > 1) {
            queue_2_.push(queue_1_.front());
            queue_1_.pop();
        }

        int answer = queue_1_.front();
        queue_1_.pop();
        while (queue_2_.size() > 0) {
            queue_1_.push(queue_2_.front());
            queue_2_.pop();
        }

        return answer;
    }

    int top() {
        int answer;
        while (!queue_1_.empty()) {
            answer = queue_1_.front();
            queue_2_.push(answer);
            queue_1_.pop();
        }

        while (queue_2_.size() > 0) {
            queue_1_.push(queue_2_.front());
            queue_2_.pop();
        }

        return answer;
    }

    bool empty() {
        return queue_1_.empty();
    }

private:
    std::queue<int> queue_1_;
    std::queue<int> queue_2_;
};

/**
 * Your MyStack object will be instantiated and called as such:
 * MyStack* obj = new MyStack();
 * obj->push(x);
 * int param_2 = obj->pop();
 * int param_3 = obj->top();
 * bool param_4 = obj->empty();
 */
